Custom types

jsamoocha opened the pull request:
closes #500 

## Description

Adds custom types to our model that allow to preserve the legacy behavior of, e.g., unit conversion, while being a proper subtype of the generated strava model and not tripping up mypy.

## Type of change

Select the statement best describes this pull request.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Does your PR include tests

If you are fixing a bug or adding a feature, we appreciate (but do not require)
tests to support whatever fix of feature you're implementing.

- [x] Yes

## Did you include your contribution to the change log?

- [x] Yes, `changelog.md` is up-to-date.

jsamoocha commented on src/stravalib/model.py:
```diff
@@ -220,6 +231,148 @@ def check_valid_location(
+    _type: Type[CustomType]
+
+    @classmethod
+    @abstractmethod
+    def validate(cls, value: BaseType) -> CustomType: ...
+
+    @classmethod
+    @abstractmethod
+    def get_core_schema(cls) -> Mapping[str, Any]: ...
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls,
+        _source_type: BaseType,  # TODO: was ANY??
+        _handler: GetCoreSchemaHandler,
+    ) -> core_schema.CoreSchema:
+        from_schema = core_schema.chain_schema(
+            [
+                cls.get_core_schema(),
+                core_schema.no_info_plain_validator_function(cls.validate),
+            ]
+        )
+
+        return core_schema.json_or_python_schema(
+            json_schema=from_schema,
+            python_schema=core_schema.union_schema(
+                [core_schema.is_instance_schema(cls._type), from_schema]
+            ),
+            serialization=core_schema.plain_serializer_function_ser_schema(
+                lambda v: v
+            ),
+        )
+
+    @classmethod
+    def __get_pydantic_json_schema__(
+        cls,
+        _core_schema: core_schema.CoreSchema,
+        handler: GetJsonSchemaHandler,
+    ) -> JsonSchemaValue:
+        return handler(cls.get_core_schema())
```
See https://docs.pydantic.dev/latest/concepts/types/#as-an-annotation. This is quite verbose, but I didn't succeed in using the simplified way explained in https://docs.pydantic.dev/latest/concepts/types/#using-getpydanticschema-to-reduce-boilerplate.

jsamoocha commented on src/stravalib/model.py:
```diff
@@ -220,6 +231,148 @@ def check_valid_location(
+        return core_schema.str_schema()
+
+
+class Duration(int):
+    def timedelta(self) -> timedelta:
+        return timedelta(seconds=self)
```
This is an example of the new extended types. We simply subclass `int` (which is the base type of, e.g., `elapsed_time` in many classes) and add an accessor method to the enhanced type.

jsamoocha replied:
We should find some way to filter out the base type inheritance from the documentation for these specific cases. I added a TODO to #502.

jsamoocha commented on src/stravalib/model.py:
```diff
@@ -220,6 +231,148 @@ def check_valid_location(
+        return Velocity(value)
+
+
+DurationType = Annotated[Duration, _DurationAnnotation]
+DistanceType = Annotated[Distance, _DistanceAnnotation]
+VelocityType = Annotated[Velocity, _VelocityAnnotation]
+TimezoneType = Annotated[Timezone, _TimezoneAnnotation]
```
These are the types that we now use in the model classes.

jsamoocha commented on src/stravalib/model.py:
```diff
@@ -398,11 +551,15 @@ class DetailedClub(SummaryClub, strava_model.DetailedClub):
 
-    pass
+    # Attribute overrides for custom types:
+    distance: Optional[DistanceType] = None
```
If `a` is an instance of `ActivityTotals`, we can now do the following:

```python
a.distance  # gets distance as a float
a.distance.quantity()  # gets distance as a Pint Quantity (in meters)
unit_helper.feet(a.distance)  # gets distance as a Pint Quantity (in feet)
```

jsamoocha replied:
That sounds good, I added #535

jsamoocha commented on src/stravalib/unithelper.py:
```diff
@@ -4,84 +4,43 @@
+    Callable that converts quantities or unitless numbers to quantities of its unit
     """
 
-    @property
-    def num(self) -> pint._typing.Magnitude:
-        warn_units_deprecated()
-        return self.magnitude
-
-    @property
-    def unit(self) -> str:
-        warn_units_deprecated()
-        return str(self.units)
-
-    def __int__(self) -> int:
-        return int(self.magnitude)
-
-    def __float__(self) -> float:
-        return float(self.magnitude)
```
We drop this backward compatibility with the `units` package we used before.

jsamoocha commented on src/stravalib/unithelper.py:
```diff
@@ -4,84 +4,43 @@
-        return converted_q
-
-
-def is_quantity_type(obj: Any) -> bool:
-    if isinstance(obj, (pint.Quantity, Quantity)):
-        return True
-    elif isinstance(obj, UnitsQuantity):  # check using Duck Typing
-        warn_units_deprecated()
-        return True
-    else:
-        return False
```
This is not needed anymore because we drop backward compatibility with the `units` package.

jsamoocha commented:
Codecov complains about dropped coverage in model.py, which I couldn't find. Maybe we can review that before mering `pydantic-v2` into `main`.

lwasser commented on src/stravalib/model.py:
```diff
@@ -220,6 +231,148 @@ def check_valid_location(
+        return Timezone(value)
+
+
+class Distance(_Quantity):
+    unit = "meters"
```
```suggestion
class Distance(_Quantity):
    """A class that supports conversion of units to meters. This output can then be 
    converted to other units such as feet or meters using the stravalib.unit_helper module. 
    
    Examples
    ----------
    Once you have distance in meters, you can then use the unit_helper module to 
    convert to other units such as feet 
    
    >>> from stravalib import unithelper
    >>> activity = client.get_activity(11416949675)
    >>> activity.distance.quantity()
    <Quantity(8055.9, 'meter')>
    >>>  unithelper.feet(activity.distance)
    <Quantity(26430.1181, 'foot')>
    >>> unithelper.miles(activity.distance)
    <Quantity(5.00570419, 'mile')>
    """
    unit = "meters"
```

lwasser commented on docs/reference/unithelper.rst:
let's rename unit-helper for consistency with other .rst files. i can also help with this

lwasser reviewed (commented):
This is awesome. 
Jonatan, I think there is an opportunity for us to document this piece of stravalib better! It's really really helpful to have these built-in conversions. Reading the docs, I am not sure all users will know about this feature! And now we are competing in our api docs with things like the docs for int (that has examples). this is not a huge deal but i just think it will help users to throw in examples throughout so they know about this functionality! 

This could be a combination of adding some examples to the docstrings + a more descriptive narrative.  
Also -  we could do that outside of this pr as well. i just added a few suggestions here and there so you could see what i'm thinking about!

lwasser commented on src/stravalib/model.py:
```diff
@@ -220,6 +231,148 @@ def check_valid_location(
+
+
+class Timezone(str):
+    def timezone(
```
Can you help me understand how this would be used? when i pull down an activity i'd get a start and end time that are in utc by default. it looks like this is checking a string. is it for updating activities or adding activities? Or could i use it to convert the start time i have from UTC --> Mountain Time ? 

An example would be really helpful!

lwasser commented on src/stravalib/model.py:
```diff
@@ -220,6 +231,148 @@ def check_valid_location(
+        return Distance(value)
+
+
+class Velocity(_Quantity):
```
Let's add a short description here that directs users to unit_helper. And we could also include a small example  using unit_helper. i'm just starting to understand how you pulled this all together - bare with me 😸

lwasser reviewed (approved):
@jsamoocha ok i think i generally understand this all now - but my big take aways here are

* let's add some quick examples in the docstrings of the new unit conversion classes
* let's add a section in our docs on unit conversions. 

i can help with this but will need a bit of guidance from you as i'm still trying to wrap my head around how this works for the various raw values that are returned. Once i understand that i can be more helpful with docs! 

i don't have any big changes except it seems like we could potentially rename unithelper.py --> unit_helper.py

jsamoocha commented on src/stravalib/model.py:
```diff
@@ -220,6 +231,148 @@ def check_valid_location(
+        return core_schema.str_schema()
+
+
+class Duration(int):
```
The suggestion is good; I like these examples! Regarding your question: the example should be

```python
activity.elapsed_time.timedelta()
```

as a method call. This will give you the timedelta object (which can be formatted with `str()`).

jsamoocha commented:
Thanks for the review! Merging this now.

I added TODOs to #502 regarding your documentation suggestions. I added #535 regarding renaming the `unithelper` module.
